### ISSUE TITLE
deal more gracefully with corrupted cache files

### ISSUE THREAD

**Issue by aquada** [borgbackup/borg#4770]
```
Mon 30 Sep 13:02:27 SAST 2019 Starting backup

Killed stale lock hornswaggle@36074249074878.3247-0.
Removed stale exclusive roster lock for host hornswaggle@36074249074878 pid 3247 thread 0.
Removed stale exclusive roster lock for host hornswaggle@36074249074878 pid 3247 thread 0.
Creating archive at "ssh://pi@192.168.0.220/media/nas/backups/hornswaggle::hornswaggle-2019-09-30T13:02:28"
Local Exception
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/borg/archiver.py", line 4501, in main
    exit_code = archiver.run(args)
  File "/usr/lib/python3.7/site-packages/borg/archiver.py", line 4433, in run
    return set_ec(func(args))
  File "/usr/lib/python3.7/site-packages/borg/archiver.py", line 166, in wrapper
    return method(self, args, repository=repository, **kwargs)
  File "/usr/lib/python3.7/site-packages/borg/archiver.py", line 566, in do_create
    cache_mode=args.files_cache_mode, ignore_inode=args.ignore_inode) as cache:
  File "/usr/lib/python3.7/site-packages/borg/cache.py", line 379, in __new__
    return local()
  File "/usr/lib/python3.7/site-packages/borg/cache.py", line 373, in local
    lock_wait=lock_wait, cache_mode=cache_mode)
  File "/usr/lib/python3.7/site-packages/borg/cache.py", line 456, in __init__
    self.open()
  File "/usr/lib/python3.7/site-packages/borg/cache.py", line 504, in open
    self.rollback()
  File "/usr/lib/python3.7/site-packages/borg/cache.py", line 609, in rollback
    self._do_open()
  File "/usr/lib/python3.7/site-packages/borg/cache.py", line 494, in _do_open
    self.chunks = ChunkIndex.read(fd)
  File "src/borg/hashindex.pyx", line 113, in borg.hashindex.IndexBase.read
  File "src/borg/hashindex.pyx", line 100, in borg.hashindex.IndexBase.__cinit__
ValueError: Incorrect file length (expected 19608574, got 4063232)

Platform: Linux hornswaggle 4.19.69-1-MANJARO #1 SMP PREEMPT Thu Aug 29 08:51:46 UTC 2019 x86_64
Linux: arch Manjaro Linux
Borg: 1.1.10  Python: CPython 3.7.4 msgpack: 0.5.6
PID: 3477  CWD: /home/daniel/bin
sys.argv: ['/usr/bin/borg', 'create', '--verbose', '--filter', 'AME', '--list', '--stats', '--show-rc', '--compression', 'zstd', '--exclude-if-present', '.noborgbackup', '--exclude-caches', '--exclude', '/home/*/.cache/*', '--exclude', '/var/cache/*', '--exclude', '/var/tmp/*', '::{hostname}-{now}', '/etc', '/home', '/root', '/var']
SSH_ORIGINAL_COMMAND: None

terminating with error status, rc 2

Mon 30 Sep 13:02:35 SAST 2019 Pruning repository

Killed stale lock hornswaggle@36074249074878.3477-0.
Removed stale exclusive roster lock for host hornswaggle@36074249074878 pid 3477 thread 0.
Removed stale exclusive roster lock for host hornswaggle@36074249074878 pid 3477 thread 0.
Local Exception
Traceback (most recent call last):
  File "/usr/lib/python3.7/site-packages/borg/archiver.py", line 4501, in main
    exit_code = archiver.run(args)
  File "/usr/lib/python3.7/site-packages/borg/archiver.py", line 4433, in run
    return set_ec(func(args))
  File "/usr/lib/python3.7/site-packages/borg/archiver.py", line 166, in wrapper
    return method(self, args, repository=repository, **kwargs)
  File "/usr/lib/python3.7/site-packages/borg/archiver.py", line 1591, in do_prune
    with Cache(repository, key, manifest, lock_wait=self.lock_wait) as cache:
  File "/usr/lib/python3.7/site-packages/borg/cache.py", line 379, in __new__
    return local()
  File "/usr/lib/python3.7/site-packages/borg/cache.py", line 373, in local
    lock_wait=lock_wait, cache_mode=cache_mode)
  File "/usr/lib/python3.7/site-packages/borg/cache.py", line 456, in __init__
    self.open()
  File "/usr/lib/python3.7/site-packages/borg/cache.py", line 504, in open
    self.rollback()
  File "/usr/lib/python3.7/site-packages/borg/cache.py", line 609, in rollback
    self._do_open()
  File "/usr/lib/python3.7/site-packages/borg/cache.py", line 494, in _do_open
    self.chunks = ChunkIndex.read(fd)
  File "src/borg/hashindex.pyx", line 113, in borg.hashindex.IndexBase.read
  File "src/borg/hashindex.pyx", line 100, in borg.hashindex.IndexBase.__cinit__
ValueError: Incorrect file length (expected 19608574, got 4063232)

Platform: Linux hornswaggle 4.19.69-1-MANJARO #1 SMP PREEMPT Thu Aug 29 08:51:46 UTC 2019 x86_64
Linux: arch Manjaro Linux
Borg: 1.1.10  Python: CPython 3.7.4 msgpack: 0.5.6
PID: 3485  CWD: /home/daniel/bin
sys.argv: ['/usr/bin/borg', 'prune', '--list', '--prefix', '{hostname}-', '--show-rc', '--keep-hourly', '4', '--keep-daily', '7', '--keep-weekly', '4', '--keep-monthly', '6']
SSH_ORIGINAL_COMMAND: None

terminating with error status, rc 2

Mon 30 Sep 13:02:41 SAST 2019 Backup and/or Prune finished with an error

```

## Have you checked borgbackup docs, FAQ, and open Github issues?
Yes

## Is this a BUG / ISSUE report or a QUESTION?
ISSUE

## System information. For client/server mode post info for both machines.
Client: Platform: Linux hornswaggle 4.19.69-1-MANJARO #1 SMP PREEMPT Thu Aug 29 08:51:46 UTC 2019 x86_64
Linux: arch Manjaro Linux
Borg: 1.1.10  Python: CPython 3.7.4 msgpack: 0.5.6

Server: Linux Debian Borg 1.1.10

#### Hardware / network configuration, and filesystems used.
Ext4
#### How much data is handled by borg?
2TB
#### Full borg commandline that lead to the problem (leave away excludes and passwords)
```
#!/bin/sh

#todo - set USERNAME variable for error messages
BORGUSER="daniel"

# Setting this, so the repo does not need to be given on the commandline:
export BORG_REPO=ssh://pi@192.168.0.220/media/nas/backups/$HOSTNAME

# Setting this, so you won't be asked for your repository passphrase:
export BORG_PASSPHRASE='XXXXXXXXXXXXXXX'
# or this to ask an external program to supply the passphrase:
#export BORG_PASSCOMMAND='pass show backup'

# some helpers and error handling:
info() { printf "\n%s %s\n\n" "$( date )" "$*" >&2; }
trap 'echo $( date ) Backup interrupted >&2; exit 2' INT TERM

info "Starting backup"
sudo -u $BORGUSER DISPLAY=:0 DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus notify-send 'Backup started!' \
'The Borg has started a backup of your computer.' --icon=dialog-information -t 14400000

# Backup the most important directories into an archive named after
# the machine this script is currently running on:

borg create                            \
    --verbose                          \
    --filter AME                       \
    --list                             \
    --stats                            \
    --show-rc                          \
    --compression zstd                 \
    --exclude-if-present .noborgbackup \
    --exclude-caches                   \
    --exclude '/home/*/.cache/*'       \
    --exclude '/var/cache/*'           \
    --exclude '/var/tmp/*'             \
    ::'{hostname}-{now}'               \
    /etc                               \
    /home                              \
    /root                              \
    /var                               \

backup_exit=$?

info "Pruning repository"

# Use the `prune` subcommand to maintain 7 daily, 4 weekly and 6 monthly
# archives of THIS machine. The '{hostname}-' prefix is very important to
# limit prune's operation to this machine's archives and not apply to
# other machines' archives also:

borg prune                          \
    --list                          \
    --prefix '{hostname}-'          \
    --show-rc                       \
    --keep-hourly   4               \
    --keep-daily    7               \
    --keep-weekly   4               \
    --keep-monthly  6               \

prune_exit=$?

# use highest exit code as global exit code
global_exit=$(( backup_exit > prune_exit ? backup_exit : prune_exit ))

if [ ${global_exit} -eq 1 ];
then
    info "Backup and/or Prune finished with a warning"
sudo -u $BORGUSER DISPLAY=:0 DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus notify-send 'BACKUP WARNING' 'The Borg has completed a backup  with a warning.' --icon=dialog-warning -t 14400000


fi

if [ ${global_exit} -gt 1 ];
then
    info "Backup and/or Prune finished with an error"
sudo -u $BORGUSER DISPLAY=:0 DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus notify-send 'BACKUP ERROR' 'The Borg has completed a backup with an error.' --icon=dialog-error -t 14400000

fi

if [ ${global_exit} -eq 0 ];
then
sudo -u $BORGUSER DISPLAY=:0 DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/1000/bus notify-send 'Backup Successful' 'The Borg has completed a backup successfully.' --icon=dialog-information -t 14400000

fi

exit ${global_exit}
```

**Comment by ThomasWaldmann** [comment]
Well, it looks like you had a crash before (that is why a stale lock existed at the time you ran the command you posted above).

Could be that this crash damaged / truncated the `chunks` file (see `~/.cache/borg/<REPOID>/chunks`).

Either remove it manually or delete the whole cache via `borg delete --cache-only REPO` (the latter will also delete all files in `chunks.archive.d/`, so it might take longer to re-sync the chunks cache at the next run).

**Comment by alleyoopster** [comment]
You guys rock. Thanks @ThomasWaldmann

**Comment by henk84** [comment]
I had a similar issue:
```
Killed stale lock frustcomp.hnjs.home.arpa@66619482969610.17428-0.
Removed stale exclusive roster lock for host frustcomp.hnjs.home.arpa@66619482969610 pid 17428 thread 0.
Removed stale exclusive roster lock for host frustcomp.hnjs.home.arpa@66619482969610 pid 17428 thread 0.
Creating archive at "ssh://henkbackup@nastix.hnjs.home.arpa/./personaldata::frustcomp-henk-personaldata-2025-03-04T08:52:34"
Local Exception
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 5213, in main
    exit_code = archiver.run(args)
                ^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 5144, in run
    return set_ec(func(args))
                  ^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 183, in wrapper
    return method(self, args, repository=repository, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 656, in do_create
    with Cache(repository, key, manifest, progress=args.progress,
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/borg/cache.py", line 387, in __new__
    return local()
           ^^^^^^^
  File "/usr/lib/python3/dist-packages/borg/cache.py", line 378, in local
    return LocalCache(repository=repository, key=key, manifest=manifest, path=path, sync=sync,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/borg/cache.py", line 487, in __init__
    self.open()
  File "/usr/lib/python3/dist-packages/borg/cache.py", line 542, in open
    self.rollback()
  File "/usr/lib/python3/dist-packages/borg/cache.py", line 662, in rollback
    self._do_open()
  File "/usr/lib/python3/dist-packages/borg/cache.py", line 527, in _do_open
    self.chunks = ChunkIndex.read(fd)
                  ^^^^^^^^^^^^^^^^^^^
  File "src/borg/hashindex.pyx", line 114, in borg.hashindex.IndexBase.read
  File "src/borg/hashindex.pyx", line 99, in borg.hashindex.IndexBase.__cinit__
ValueError: Incorrect file length (expected 11534486, got 4194304)

Platform: Linux frustcomp 6.1.0-28-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.119-1 (2024-11-22) x86_64
Linux: Unknown Linux
Borg: 1.2.4  Python: CPython 3.11.2 msgpack: 1.0.3 fuse: None [pyfuse3,llfuse]
PID: 3720  CWD: /home/henk
sys.argv: ['/usr/bin/borg', 'create', '-v', '--lock-wait', '600', '--one-file-system', '--compression', 'auto,zstd', '--patterns-from=/home/henk/.backup/patterns/personaldata', '::{hostname}-{user}-personaldata-{now}']
SSH_ORIGINAL_COMMAND: None
```

After moving the chunksfile, the error changed:
```
Killed stale lock frustcomp.hnjs.home.arpa@66619482969610.8219-0.
Removed stale exclusive roster lock for host frustcomp.hnjs.home.arpa@66619482969610 pid 8219 thread 0.
Removed stale exclusive roster lock for host frustcomp.hnjs.home.arpa@66619482969610 pid 8219 thread 0.
Creating archive at "ssh://henkbackup@nastix.hnjs.home.arpa/./personaldata::frustcomp-henk-personaldata-2025-03-04T09:05:42"
Local Exception
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 5213, in main
    exit_code = archiver.run(args)
                ^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 5144, in run
    return set_ec(func(args))
                  ^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 183, in wrapper
    return method(self, args, repository=repository, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 656, in do_create
    with Cache(repository, key, manifest, progress=args.progress,
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/borg/cache.py", line 387, in __new__
    return local()
           ^^^^^^^
  File "/usr/lib/python3/dist-packages/borg/cache.py", line 378, in local
    return LocalCache(repository=repository, key=key, manifest=manifest, path=path, sync=sync,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/borg/cache.py", line 487, in __init__
    self.open()
  File "/usr/lib/python3/dist-packages/borg/cache.py", line 542, in open
    self.rollback()
  File "/usr/lib/python3/dist-packages/borg/cache.py", line 662, in rollback
    self._do_open()
  File "/usr/lib/python3/dist-packages/borg/cache.py", line 525, in _do_open
    with IntegrityCheckedFile(path=os.path.join(self.path, 'chunks'), write=False,
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3/dist-packages/borg/crypto/file_integrity.py", line 129, in __init__
    self.file_fd = override_fd or open(path, mode)
                                  ^^^^^^^^^^^^^^^^
FileNotFoundError: [Errno 2] No such file or directory: '/home/henk/.cache/borg/d17a6ddfced683aefd0a59018faf5664fb822a630c6394966658ebfa38e8b2e9/chunks'

Platform: Linux frustcomp 6.1.0-28-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.119-1 (2024-11-22) x86_64
Linux: Unknown Linux
Borg: 1.2.4  Python: CPython 3.11.2 msgpack: 1.0.3 fuse: None [pyfuse3,llfuse]
PID: 8963  CWD: /home/henk
sys.argv: ['/usr/bin/borg', 'create', '-v', '--lock-wait', '600', '--one-file-system', '--compression', 'auto,zstd', '--patterns-from=/home/henk/.backup/patterns/personaldata', '::{hostname}-{user}-personaldata-{now}']
SSH_ORIGINAL_COMMAND: None
```

After then running `borg delete --cache-only`, borg rebuilt the cache.

IMHO the error handling on this should be improved so borgbackup fails a bit more gracefully and with more helpful error messages.

**Comment by ThomasWaldmann** [comment]
@henk84 yeah, it is a bit unpretty, but the root cause is likely that your system or borg is crashed in the run before. That's also the reason for the stale lockfile it had to remove.

**Comment by henk84** [comment]
> [@henk84](https://github.com/henk84) yeah, it is a bit unpretty, but the root cause is likely that your system or borg is crashed in the run before. That's also the reason for the stale lockfile it had to remove.

Yup, that’s correct, that happened. And it’s just a reality that this can happen, so it should be dealt with gracefully IMHO.
I’m not sure whether your response implies that it will not be improved because the root cause was a crash (I’m a bit impaired when it comes to understanding implications). I’d find that reasoning a bit odd, to be honest.
I would recommend reopening this issue and changing the title to something like "Improve exception handling in case of corrupted cache", to make a concrete suggestion for further steps. What do you think?

**Comment by ThomasWaldmann** [comment]
PRs welcome!

**Comment by ThomasWaldmann** [comment]
@henk84 I just noticed the original bug report is with a meanwhile unsupported borg version and your issue also was with a rather old version.

So, I'ld suggest you first try to reproduce with a current borg version, like 1.2.8 or 1.4.0 and if it still happens, we'll reopen this.

**Comment by ThomasWaldmann** [comment]
Note: Might be fixed by the fix of #5213, which is in 1.4-maint branch and will be released with 1.4.1.

**Comment by maxboehm** [comment]
Note for anyone encountering this issue:

This exception concerns client-side cache data. Run `borg delete --cache [REPO_PATH]` -only **on the machine where the error occurs**—it can be the server, but doesn’t have to be.

**Comment by jtmackoy** [comment]
tl;dr for Vorta installed via Flatpak:
`flatpak run --command=borg com.borgbase.Vorta delete --cache <path-to-repo>`

For those who have encountered this in the context of another client that invokes borg, it's useful to remember that the client-side cache should be cleared in the same context your specific client also invokes borg.

Also mentioned [in the Vorta repo](https://github.com/borgbase/vorta/issues/2358#issuecomment-4215179193).
